pkg/apitype/stack_references.go

```diff
@@ -0,0 +1,37 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apitype
+
+// StackReferenceRequest creates or updates a stack reference to the specified stack.
+type StackReferenceRequest struct {
+	// OrgName is the organization name the stack being referenced is found in.
+	OrgName string `json:"orgName"`
+	// ProjectName is the name of the project the stack being referenced is associated with.
+	ProjectName string `json:"projectName"`
+	// StackName is the name of the stack being referenced.
+	StackName string `json:"stackName"`
+	// Outputs that the stack references.
+	Outputs []string `json:"outputs"`
```

Comment:
Without changes to the user-facing `StackReference` API, I don't think we can know this information.

Reply:
What is the current user-facing `StackReference` API? Is there a specific endpoint on https://api.pulumi.com that serves this information? My understanding was that the stack reference code within the CLI was just calling `/export` on the stack, and then looking for the stack outputs from the result.

So by adding this first class support, we could create a new `StackReference` API that just exposed the outputs (and not the rest of the checkpoint, for example).

Looking through the CLI code, I didn't quite see what APIs were called as part of stack references. Is the code for how the engine creates stack references in `pkg/resource/deploy/builtins.go`? I couldn't quickly determine how things eventually mapped to the `httpstate` backend, and translating to an API call.

Reply:
https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/stackReference.ts#L19-L123

A stack reference is currently constructed using the name of the referenced stack. The implementation calls `GetStackOutputs` on the backend interface, which ends up calling `/export` and finding the stack's outputs from the state file. After the stack reference is created, the user calls `getOutput` to get specific outputs.

```
// Create a reference to stack "foo". The reference will have all of foo's outputs.
var foo = new StackReference("foo");

// Use specific outputs from "foo".
foo.getOutput("a").apply(a => console.log(a));
```

In order to have information about output usage, we would need to know the outputs the user needed at the point of `new StackReference`. @ekrengel can you expand on what in https://github.com/pulumi/pulumi/issues/2309 seems like it could use this info?

Reply:
> Warnings/errors if you update a stack output that is being depended on, letting you know which other stacks you will need to update.

A stack can have many outputs, but only some may be referenced. It would be good to know which are being referenced so we can:

1. warn you that your updating a value that is referenced by another stack
1. let downstream users know the upstream value they care about has been updated

That being said, I'll go ahead and remove it for now. It will be a nice feature w/o the specific details of which outputs are relied on. We can always add it later.

Reply:
For posterity, I picked Pat's brain this morning to learn more about how stack references worked in the CLI.

Another useful piece of code is here:
https://github.com/pulumi/pulumi/blob/master/pkg/resource/deploy/builtins.go#L129

This is how the CLI handles requests to create/update/delete stack reference resources. The linked code in `sdk/nodejs/stackReference.ts` is the bits from `@pulumi/pulumi`. But the logic where we actually contact the Pulumi Service (via `pkg/backend`) to read the stack's outputs is in the "builtin" resource provider.